src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+    NonSiege, // Castle
+    Villagers,  // Khmer houses
+    /// Units from production buildings can be trained directly into garrison,
+    /// but not garrison from outside.
+    Production,
+    /// Same as Production plus Relics can garrison
+    ProductionAndRelics
+  };
+
+  GarrisonType garrisonType;
+  int garrisonCapacity;
+
+  /// Based on the type:
+  /// - Villager, FishingShip: gather rate
+  /// - Production building: production speed
+  /// - Other: ?
+  float workRate;
+
+  int conversionResistanceLevel;
+
+  // Used only by resource spot and animals
+  ResourceAmount resources;
+};
+
+constexpr int sizeof_ObjectTypeStats = sizeof(ObjectTypeStats); 
+// 320 = 2 * sizeof(Damage) + 80
+// 200 types * 2 per info * 320 = 125K
+
+struct UnitTypeStats : public ObjectTypeStats {
+
+  float radius;
+  float speed;
+};
+
+struct BuildingTypeStats : public ObjectTypeStats {
+
+  int populationSpace;
+  bool dropOffPoint[4];
+
+  QSize size;
+  QRect occupancy;
+
+};
+
+
+/// CivilizationStats contain every info that applies in a global level to all the units and
+/// buildings.
+///
+/// CRITICAL: Also here are all the information about values that are needed for a single unit.
+///
+///           Example: villagerCarryingCapacity: an obvius design would be to add this
+///           to the villager unit type that will contain this. Because only the Villager need
+///           this, it would introduce some variation in the UnitTypeStats, like inheritance or
+///           a list of generic attribute objects. The other solution is to put in the CivilizationStats.
+///
+///           Problems solved by keeping the UnitTypeStats as simple struct without pointers and lists.
+///           - Can be placed inside a vector<UnitTypeStats> (without pointer) = memory management done
+///           - The number of UnitTypeStats is known at start + a properly sized vector =
+///             simple pointers to UnitTypeStats can be used safely
+///           - Huge reduction of the code base.
+///           - Cleaner implemention of the gamelogic.
+///           - Simpler information retrieval for the gui.
+///           - Performance (size, locality, less indirection)
+///
+///
+struct CivilizationStats {
+
+  // Add to the starting resource of the player
+  ResourceAmount startingResourceD;
+
+  UnitType startingScoutUnit;
+  int startingVillagerCount;
+  bool startingLlama;
+
+  int populationCap;
+  int bonusPopulationSpace;
+
+  float garrisonHealRate;
+
+  // TODO: monk conversion things
+
+  float tributeFee;
+  float tradingFree;
+  float tradeGoldGeneration;
+  float relicGoldGeneration;
+
+  // == single unit values
+
+  int farmFood;
+
+  // NOTE: could aslo be stored as the workRate of an unspecialized villager
+  float villagerConstuctionSpeed;
+
+  float villagerCarryingCapacity[/* TODO: number of villager types*/];
+  float fishingShipCarryingCapacity;
+
+  bool singleTowerCenter;
```

Comment:
Switch in to an `int maxTownCenters` should do the trick.

> constructing it in this age take longer than usual

This could be implemented like this:
```c++
  age[0].addModification( // increase on game start
    FilterOfBuildingsByType(BuildingType::TownCenter),
    Modification(Type::ConstructionSpeed, Operation::Add, X) // I cannot find the value in the wiki...
  )
  age[2].addModification( // decrease on castle age to return to original value
    FilterOfBuildingsByType(BuildingType::TownCenter),
    Modification(Type::ConstructionSpeed, Operation::Add, -X)
  )
```
Other subject: If this scenario come up in many places there could be two modification list per age:
- modifications that apply from this age until the end
- modifications that apply to only this age

(and other two for the team bonus)

Reply:
I am not sure, but it seems likely that the slower construction only applies to the second town center, so this might necessitate an extra setting in order to keep construction of the first one at the standard speed.

Reply:
I found this in the wiki:

>  For any Town Center built in the Feudal Age, the construction time is increased by 80% (2:30 → 4:30 minutes for a single Villager construction time); this includes the situation where the player has no Town Center. The debuff is removed instantly upon reaching the Castle Age, even for a Town Center which foundation was laid in the Feudal Age.

I will believe this because it causes less problems :) . And I believe the last sentence is how the the current code would behave.

Reply:
Okay, sorry for suspecting the wrong thing then :) This behavior seems a bit weird, but indeed makes things easier at least.